iothub/device/src/Transport/AmqpIoT/AmqpIoTReceivingLink.cs

```diff
@@ -0,0 +1,219 @@
+        internal void RegisterTwinListener(Action<Twin, string, TwinCollection> onDesiredPropertyReceived)
+        {
+            _onDesiredPropertyReceived = onDesiredPropertyReceived;
+            _receivingAmqpLink.RegisterMessageListener(OnDesiredPropertyReceived);
+        }
+
+        private void OnDesiredPropertyReceived(AmqpMessage amqpMessage)
+        {
+            if (Logging.IsEnabled) Logging.Enter(this, amqpMessage, $"{nameof(OnDesiredPropertyReceived)}");
+            try
+            {
+                _receivingAmqpLink.DisposeDelivery(amqpMessage, true, AmqpIoTConstants.AcceptedOutcome);
+                string correlationId = amqpMessage.Properties?.CorrelationId?.ToString();
+
+                if (!VerifyResponseMessage(amqpMessage))
+                {
+                    _onDesiredPropertyReceived.Invoke(null, correlationId, null);
+                }
+
+
+                Twin twin = null;
+                TwinCollection twinProperties = null;
+
+                if (correlationId != null)
+                {
+                    if (amqpMessage.BodyStream != null)
+                    {
+                        // This a result of a GET TWIN so return (set) the full twin
+                        using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                        {
+                            string body = reader.ReadToEnd();
+                            var properties = JsonConvert.DeserializeObject<TwinProperties>(body);
+                            twin = new Twin(properties);
+                        }
+                    }
+                    else
+                    {
+                        // This is a desired property ack from the service
+                        twin = new Twin();
+                    }
+                }
+                else
+                {
+                    // No correlationId, this is a PATCH sent by the sevice so return (set) the TwinCollection
+
+                    using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                    {
+                        string patch = reader.ReadToEnd();
+                        twinProperties = JsonConvert.DeserializeObject<TwinCollection>(patch);
+                    }
+                }
+                _onDesiredPropertyReceived.Invoke(twin, correlationId, twinProperties);
+            }
+            finally
+            {
+                if (Logging.IsEnabled) Logging.Exit(this, amqpMessage, $"{nameof(OnDesiredPropertyReceived)}");
+            }
+        }
+        #endregion
+
+        internal static bool VerifyResponseMessage(AmqpMessage response)
+        {
+            bool retVal = true;
+            if (response != null)
+            {
+                if (response.MessageAnnotations.Map.TryGetValue(AmqpIoTConstants.ResponseStatusName, out int status))
+                {
+                    if (status >= 400)
```

Comment:
Shouldn't all these be stored as constants somewhere?

Reply:
Agree, but won't fix for now, it should be done across all the protocols. Later...

---
In reply to: [293022001](https://github.com/Azure/azure-iot-sdk-csharp/pull/947#discussion_r293022001) [](ancestors = 293022001)